array.go

```diff
@@ -1197,6 +1197,253 @@ func (a *Array) NotContainsAny(values ...interface{}) *Array {
+				Expected:  &AssertionValue{y.value},
+				Reference: &AssertionValue{a.value},
+				Errors: []error{
+					errors.New("expected: array is ordered"),
+					fmt.Errorf("element %v must be less than element %v, but it isn't",
+						x.value, y.value),
+				},
+			})
+			return a
+		}
+	}
+
+	return a
+}
+
+// NotOrdered succeeds if array is not ordered based on optional `comparator` function.
+// For default, it will use built-in comparator function for each data type.
+// Built-in comparator requires all elements in the array to have same data type.
+//
+// Example:
+//
+//	array := NewArray(t, []interface{}{102, 101, 100})
+//	array.NotOrdered() // succeeds
+//	array.IsOrdered(func(x, y *httpexpect.Value) bool {
+//		return x.Number().Raw() < y.Number().Raw()
+//	}) // succeeds
+func (a *Array) NotOrdered(less ...func(x, y *Value) bool) *Array {
+	a.chain.enter("NotOrdered()")
+	defer a.chain.leave()
+
+	if a.chain.failed() {
+		return a
+	}
+
+	if len(less) > 1 {
+		a.chain.fail(AssertionFailure{
+			Type: AssertUsage,
+			Errors: []error{
+				errors.New("unexpected multiple less arguments"),
+			},
+		})
+		return a
+	}
+
+	if len(a.value) <= 1 {
+		a.chain.fail(AssertionFailure{
+			Type:      AssertLt,
+			Actual:    &AssertionValue{a.value},
+			Expected:  &AssertionValue{a.value}, // FIXME: reverse
+			Reference: &AssertionValue{a.value},
+			Errors: []error{
+				errors.New("expected: array is not ordered"),
+				// fmt.Errorf("element %v must not be less than element %v, but it is",
+				// 	x.value, y.value),
+			},
+		})
+		return a
+	}
+
+	var lessFn func(x, y *Value) bool
+	if len(less) == 1 {
+		lessFn = less[0]
+	} else {
+		lessFn = a.getBuiltinComparator()
+		if a.chain.failed() {
+			return a
+		}
+	}
+
+	var i int
+	ordered := true
+	for i = 0; i < len(a.value)-1; i++ {
+		xChain := a.chain.clone()
+		xChain.replace("NotOrdered[%d]", i)
+		x := newValue(xChain, a.value[i])
+
+		yChain := a.chain.clone()
+		yChain.replace("NotOrdered[%d]", i+1)
+		y := newValue(yChain, a.value[i+1])
+
+		if lessFn(y, x) {
+			ordered = false
+			break
+		}
+	}
+
+	if ordered {
+		a.chain.fail(AssertionFailure{
+			Type:      AssertLt,
+			Actual:    &AssertionValue{a.value},
+			Expected:  &AssertionValue{a.value}, // FIXME:
+			Reference: &AssertionValue{a.value},
+			Errors: []error{
+				errors.New("expected: array is not ordered"),
+				// fmt.Errorf("element %v must not be less than element %v, but it is",
+				// 	x.value, y.value),
+			},
+		})
+	}
+
+	return a
+}
+
+func (a *Array) getBuiltinComparator() func(x, y *Value) bool {
+	var fn func(x, y *Value) bool
+
+	a.validateElementsType()
+	if a.chain.failed() {
+		return fn
+	}
+
+	fn = a.constructLessFn()
+
+	return fn
+}
+
+func (a *Array) validateElementsType() {
+	var prev interface{}
+	for idx, curr := range a.value {
+		switch curr.(type) {
+		case bool, float64, string, nil:
+			// do nothing
+			// for better performance, function is generated after validating all elements
```

Comment:
I'd replace this:

```
// do nothing
// for better performance, function is generated after validating all elements
```

with just this:

```
// ok, do nothing
```

After splitting the code into functions, it's now obvious without comment that construction happens after the loop.